src/GraphQL/Instrumentation/TelemetryProvider.cs

```diff
@@ -0,0 +1,135 @@
+/// <para>
+/// To use a derived class, call <see cref="GraphQLBuilderExtensions.ConfigureExecution{TConfigureExecution}(IGraphQLBuilder)"/> with the type of your derived class.
+/// </para>
+/// </summary>
+public class TelemetryProvider : IConfigureExecution
+{
+    private readonly ActivitySource _activitySource;
+    private readonly TelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL.NET";
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public TelemetryProvider(TelemetryOptions options)
+    {
+        _activitySource = new ActivitySource(SourceName, typeof(TelemetryProvider).Assembly.GetNuGetVersion());
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        // start the Activity
+        using var activity = await StartActivity(options).ConfigureAwait(false);
+
+        // if no event listeners, do not record any telemetry
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTags(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
+
+        // execute the request
+        var result = await next(options).ConfigureAwait(false);
+
+        // record the status
+        await SetResultTags(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivity(ExecutionOptions options)
+        => new(_activitySource.StartActivity(ACTIVITY_OPERATION_NAME));
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="ExecutionOptions"/>.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.document</c> tags.
+    /// </summary>
+    protected virtual Task SetInitialTags(Activity activity, ExecutionOptions options)
+    {
+        activity.SetTag("graphql.operation.name", options.OperationName);
+        if (_telemetryOptions.RecordDocument && activity.IsAllDataRequested)
+            activity.SetTag("graphql.document", options.Query);
```

Comment:
I am not familiar with the code base, but if there is a sanitization method available it could be used here

`[1]: The value may be sanitized to exclude sensitive information.`

Reply:
Do you mean sensitive data in `options.Query`?

Reply:
Yes, if there is any chance that the query can contain sensitive data.

Reply:
It is possible to scrub the query for literal values...although it would take some CPU power and memory to do so.  `IsAllDataRequested` as such is then useful.

Reply:
One more argument to move that code into delegate in options class along with sanitization method.

Reply:
I most certainly would not move the code into the delegate; the delegates should have no default behavior.  See my comments elsewhere on this.  Either we can introduce a property or users can derive from the class.  Besides, such a behavior should be provided by the infrastructure, if it is deemed necessary, as it will be complex for users to properly implement.

A property might be:

```cs
public bool ScrubLiterals { get; set; }
```

or

```cs
public RecordDocumentOptions RecordDocument { get; set; }

public enum RecordDocumentOptions
{
    None,
    Standard,
    ScrubLiterals,
}
```

(naming is bad, but just for example)

Reply:
> the delegates should have no default behavior.

OK